### PR TITLE
First pass at removing `set-output`

### PR DESCRIPTION
**Pull request by desrosj**
Work in progress. But feedback is welcome.

## Description
`set-output` has been deprecated. See https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/.

## Motivation and context
This will remove `deprecated` notices in workflow summaries.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by jrfnl** [comment]
*grin* @desrosj Only now seeing your PR, I was working on the same thing ;-)

File: tests/expect/composer.lock [hunk 1]
```diff
@@ -64,5 +64,5 @@
     "prefer-lowest": false,
     "platform": [],
     "platform-dev": [],
-    "plugin-api-version": "2.2.0"
+    "plugin-api-version": "2.0.0"
```
Comment (jrfnl) [inline tests/expect/composer.lock]: I have a feeling this change doesn't belong in this PR ?

File: tests/fixtures/with-lock-file/composer.lock [hunk 2]
```diff
@@ -8,20 +8,24 @@
     "packages": [
         {
             "name": "ehime/hello-world",
```
Comment (jrfnl) [inline tests/fixtures/with-lock-file/composer.lock]: The changes to this lock file need to be undone as this is explicitly testing that a non-latest locked version of the `hello-world` package will be downloaded, so updating this lock file means that's no longer being tested.

**Review by jrfnl — CHANGES_REQUESTED** [review]
While I haven't tested this, the code changes look good (save for the `composer.lock` file changes which need to be undone).

File: bin/composer_paths.sh [hunk 3]
```diff
@@ -53,7 +53,7 @@ echo "::debug::${composer_version}"
 echo "::debug::Composer cache directory found at '${cache_dir}'"
 echo "::debug::File composer.json found at '${composer_json}'"
 echo "::debug::File composer.lock path computed as '${composer_lock}'"
-echo "::set-output name=command::${composer_path}"
-echo "::set-output name=cache-dir::${cache_dir}"
-echo "::set-output name=json::${composer_json}"
-echo "::set-output name=lock::${composer_lock}"
+echo "command=${composer_path}" >> "${GITHUB_OUTPUT}"
```
Comment (jrfnl) [inline bin/composer_paths.sh]: Shellcheck doesn't like this output being called `command`...

```
In bin/composer_paths.sh line 56:
echo "command=${composer_path}" >> "${GITHUB_OUTPUT}"
^-- SC2129: Consider using { cmd1; cmd2; } >> file instead of individual redirects.

For more information:
  https://www.shellcheck.net/wiki/SC2129 -- Consider using { cmd1; cmd2; } >>...
```

File: tests/expect/composer_paths_01.exp [hunk 4]
```diff
@@ -9,9 +12,24 @@ expect "::debug::Composer path is '*'\r
...
+set fileData [read $fp]
+close $fp
+
+set expectedValue "command=*\r
+cache-dir=*\r
+json=./composer.json\r
+lock=./composer.lock\n
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/composer_paths_01.exp:29]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/composer_paths_05.exp [hunk 5]
```diff
@@ -10,9 +13,24 @@ expect "::debug::Unable to find composer.lock at '../fixtures/no-lock-file/compo
...
+set fileData [read $fp]
+close $fp
+
+set expectedValue "command=*\r
+cache-dir=*\r
+json=../fixtures/no-lock-file/composer.json\r
+lock=\r
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/composer_paths_05.exp:31]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/composer_paths_06.exp [hunk 6]
```diff
@@ -9,9 +12,24 @@ expect "::debug::Composer path is '*'\r
...
+set fileData [read $fp]
+close $fp
+
+set expectedValue "command=*\r
+cache-dir=*\r
+json=../fixtures/with-lock-file/composer.json\r
+lock=../fixtures/with-lock-file/composer.lock\r
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/composer_paths_06.exp:30]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/composer_paths_08.exp [hunk 7]
```diff
@@ -9,9 +12,24 @@ expect "::debug::Composer path is '*'\r
...
+set fileData [read $fp]
+close $fp
+
+set expectedValue "command=*\r
+cache-dir=*\r
+json=../fixtures/out-of-sync-lock/composer.json\r
+lock=../fixtures/out-of-sync-lock/composer.lock\n
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/composer_paths_08.exp:30]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/composer_paths_09.exp [hunk 8]
```diff
@@ -9,9 +12,24 @@ expect "::debug::Composer path is '../fixtures/composer.phar'\r
...
+set fileData [read $fp]
+close $fp
+
+set expectedValue "command=../fixtures/composer.phar\r
+cache-dir=*\r
+json=./composer.json\r
+lock=./composer.lock\r
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/composer_paths_09.exp:30]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/php_version_01.exp [hunk 9]
```diff
@@ -1,12 +1,31 @@
...
+set fp [open $gitHubOutputFile r]
+set fileData [read $fp]
+close $fp
+
+set expectedValue "path=*\r
+version=*.*.*\r
+\n
+"
+
+if { $expectedValue != $fileData } {
```
Comment (jrfnl) [inline tests/expect/php_version_01.exp:26]: This test fails as it needs to do a regex compare, not a direct comparison.

File: tests/expect/composer.lock (hunk 1 above)
Comment (desrosj) [reply tests/expect/composer.lock]: Yep, not sure how this was able to sneak in.

File: tests/expect/composer_paths_01.exp (hunk 4 above)
Comment (desrosj) [reply tests/expect/composer_paths_01.exp:29]: My bash scripting is admittedly poor. Do you have an example I could use as a guide?

File: bin/composer_paths.sh (hunk 3 above)
Comment (desrosj) [reply bin/composer_paths.sh]: I've changed the variable name to `composer_command`.

It doesn't seem like this will have any backwards compatibility concerns since it's only used within the repo. But not sure if I'm missing something.

**Comment by desrosj** [comment]
I've addressed a handful of the feedback given. I also added some brief inline documentation to the tests.

File: bin/composer_paths.sh (hunk 3 above)
Comment (jrfnl) [reply bin/composer_paths.sh]: I'd need to check, but this output _may_ also be available to steps in external workflows being run _after_ the step using this action runner.
If that's the case and while it would be rare for a workflow to use the `command` output, it would still need a changelog entry. Though as I said, this would need to be checked and confirmed first.

File: tests/expect/composer_paths_01.exp (hunk 4 above)
Comment (jrfnl) [reply tests/expect/composer_paths_01.exp:29]: My bash scripting skills are even worse, then again, my Google skills are reasonable ;-)

* Looks like a [wildcard compare is possible](https://stackoverflow.com/a/19897118), but the pattern used here is probably too complex for that.
* This [guide about pattern matching and regex in bash](https://riptutorial.com/bash/topic/3795/pattern-matching-and-regular-expressions) looks very helpful.
* And [this one](https://www.shell-tips.com/bash/wildcards-globbing/#gsc.tab=0) may be of use as well.

File: tests/expect/cache_key_01.exp [hunk 10]
```diff
@@ -27,5 +33,19 @@ if { $expectedValue != $fileData } {
     exit 1
 }
 
+# Confirm environment variables.
```
Comment (jrfnl) [inline tests/expect/cache_key_01.exp:36]:
```suggestion
# Confirm output variables.
```

(here and in the other test files)

**Comment by jrfnl** [comment]
@ramsey Would you happen to be able to give some guidance ? Would be nice if the CI related PRs (this one and #237) could get merged soonish, so I can pull the PR for the feature request in #234

**Comment by ramsey** [comment]
Thank you! I've been away for a bit, but I'll take a look at this today.

**Comment by desrosj** [comment]
I won't have a chance to circle back to this until Wednesday. @ramsey feel free to make any changes to wrap this up if you'd like to get it out sooner!

**Comment by ramsey** [comment]
I saw the build errors in #237 but assumed they'd go away when merged into this PR. I'll try to find some time this week to see if I can help resolve things to get this merged and tagged.

Thanks!

**Comment by jrfnl** [comment]
> I saw the build errors in #237 but assumed they'd go away when merged into this PR. I'll try to find some time this week to see if I can help resolve things to get this merged and tagged.
> 

They will (should) once the strict comparisons have been changed to regex/wildcard-based comparisons for the tests I flagged up above ;-)

**Comment by jrfnl** [comment]
Anything I can do to help move this forward ?

**Comment by desrosj** [comment]
> Anything I can do to help move this forward ?

I haven't had a chance to circle back to this because of a few other things going on at the moment. I'll likely have time next week, but if you'd like to figure out the needed test adjustments, I'm more than happy to hand the rest of this over to someone else in the interest of seeing this fixed/released.

**Comment by jrfnl** [comment]
> > Anything I can do to help move this forward ?
> 
> I haven't had a chance to circle back to this because of a few other things going on at the moment. I'll likely have time next week, but if you'd like to figure out the needed test adjustments, I'm more than happy to hand the rest of this over to someone else in the interest of seeing this fixed/released.

Okay, so I've just spend a couple of hours trying to get this to work and in the end I ended up in the [Tcl manual](https://www.tcl.tk/man/tcl8.5/) and it looks like I may have actually managed to get it working... I need to run some more tests to be sure, but I'm hoping to send in a PR later today combining @desrosj commits from this PR with the fix I found.

**Comment by jrfnl** [comment]
Wowie! Tested (including deliberately breaking tests to make sure the solution will correctly fail tests when needed) and I now have a 🟢 build. Will clean up my WIP branch and pull it momentarily.

**Comment by jrfnl** [comment]
I've opened PR #238 which builds onto this one and gets us to a green CI build again.
